### ISSUE TITLE
Migration of applications to new infrastructure

### ISSUE THREAD

**Issue by mwinokan** [m2ms/fragalysis-frontend#1451]
To update to new kubernetes versions, etc the deployments will have to be moved to new hardware infrastructure.

@tdudgeon estimates it will be at least nine days to set up the new hardware, test deployments, and then migrate the application/database.

The main driver is to improve stack resilience (and not performance). @tdudgeon please itemise the likely issues that this migration would solve. @phraenquex to make a decision based on those issues whether this is a priority on the current POs.

**Comment by alanbchristie** [comment]
Crucial components that would benefit, especially relating to our ongoing ROFS errors: -

1. **A recent longhorn**. We know longhorn has a "feature" that is documented to help under the ROFS issues. Whether it will be successful or not is not clear, but at the moment we are unable to use this version of longhorn (which relies on a more recent kubernetes).
2. **Potentially fix the prometheus alert "noise"** where we are unable to control noisy alerts.

> It must be stressed that it is not 100% clear that the above will fix our issue but it should give us additional mechanisms to automatically deal with it.

There are addition side-benefits of an upgrade: -

- More recent components, like: -
  - **Rancher** (which configures and controls our clusters)
  - Pod **startup probes** (available in more recent kubernetes) that potentially allow for faster deployments
- The use of a supported **kubernetes** (1.23 is no longer supported)

**Comment by alanbchristie** [comment]
A new Rancher (**v2.9.2**) has been installed (on the `new-world` VM in the xchem-follow-up project). It has been used to create a one-node cluster (etcd + control + worker) for Kubentetes **1.30** (a very recent release). This indicates that creating a new (post 1.25) kubernetes cluster appears to be supported on the VMs we have.

Lack of compute resources prevents us from creating a useful cluster and so the cluster has now been removed (Rancher remains).

We will produce a checklist of migration tasks along with the approximate expected effort.

**Comment by mwinokan** [comment]
@alanbchristie is testing on another cluster in parallel

**Comment by alanbchristie** [comment]
There is still a lot to do but the new developer cluster, which is using kubernetes 1.30, but it should be ready for testing developer stacks now. There may be some small issues to fix but Kalev and I have deployed stacks and logged in to CAS.

Main changes so far: -

1. **keycloak v23**. Keycloak has been upgraded and its database has be imported. It’s quite a visual change from an administrative perspective but from a service standpoint nothing should have changed.
2. **AWX v24**. The developer AWX server has been upgraded. Oh, and it is on a new URL (https://awx.xchem-dev.diamond.ac.uk/). Logins have been preserved but all the projects and templates have had to be re-created.
3. **PostgreSQL v15.8**. The latest fragalysis stack playbooks now install PostgreSQL v15.8 in order to align with other databases we’re using (we were using v12).
4. **Redis v7.4** The underlying redis service has also been upgraded (from 7.2 to 7.4)

The new cluster also brings new **longhorn**, **NFS** and **cinder** storage provisioners - so we _might_ see less disruption on Monday mornings?

In summary (so far) ... a very new kubernetes, and a leap forward for AWX, PostgreSQL and keycloak. Otherwise the stack development remains unchanged.

> I can supply an updated `KUBECONFIG` to anyone who needs one.

**Comment by mwinokan** [comment]
@alanbchristie says he is ready to change the public facing URLs to the new dev cluster, but a decision needs to be made on whether Alan does the migration or we start the dev clusters with empty stacks.

@alanbchristie will deploy some new stacks ahead of the weekend to test the ROFS errors on Sunday

**Comment by mwinokan** [comment]
@alanbchristie says that the dev stacks are all operational and have the same functionality as before (same URLs). The new cluster did exhibit some ROFS errors but they did resolve themselves.

**Comment by mwinokan** [comment]
@alanbchristie says he will need 1-2 more days to finish the migration of squonk and verify that backups are working. Then next week the migration of the production cluster can be tackled.

There is an issue of resources as dev resources on the old cluster are also being used for the fragmentation #1486

**Comment by mwinokan** [comment]
@alanbchristie has a new cluster ready to go for the production stack, but to move the applications (keycloak, stacks) the IP addresses need to be switched to the new URLs

**Comment by alanbchristie** [comment]
# Pre-production migration information

> Some production cluster migration preparation notes.
> Basically personal notes to form the "menu" and execution order for the migration.

The destination cluster exists and contains all the essential "infrastructure" components: -

- Nginx
- Certificate manager
- Cinder provisioner
- NFS provisioner
- AWX installation (with graph deployed)
- Keycloak (DB import required)

> No longhorn. Longhorn will not be used for any of the deployed application volumes during the migration.

## Existing stack versions/AWX config (k8s 1.23): -

Legacy Production
  - DB: docker.io/postgres:12.2
  - STACK: docker.io/xchem/fragalysis-stack:2023.11.1-1
  - BACKUP: informaticsmatters/sql-backup:stable

V2 Production
  - DB: docker.io/library/postgres:12.16
  - STACK: docker.io/xchem/fragalysis-stack:2024.11.3
  - BACKUP: informaticsmatters/sql-backup:stable

STAGING
  - DB: docker.io/library/postgres:12.16
  - STACK: docker.io/xchem/fragalysis-stack:latest
  - BACKUP: -

AWX project versions: -

Legacy Production
  - Template: Production Fragalysis Stack (Legacy)
  - Project: fragalysis-stack-kubernetes (v1-stack)
  - Tag: v1-stack

V2 Production
  - Template: Production Fragalysis Stack
  - Project: fragalysis-stack-kubernetes (2024.16)
  - Tag: tags/2024.16

Staging
  - Template: Staging Fragalysis Stack
  - Project: fragalysis-stack-kubernetes (2024.16)
  - Tag: tags/2024.16

CI/CD Users
- github-action

## Migration steps (approximate order)

Update MOTD message on V2/Staging stacks

Configure remaining AWX replacement objects
- Credentials
- Users
- Projects
- Job templates

Patch GRAPH

Move keycloak DB

Deploy PROD stack (production-stack) [legacy]
Deploy PROD-V2 stack (v2-production-stack)
Deploy STAGING stack (staging-stack)

Backup/restore PROD DB (gizmo volume replication)
Backup/restore PROD MEDIA (153Gi) (gizmo volume replication)
Backup/restore PROD-V2 DB (gizmo copy) (gizmo volume replication)
Backup/restore PROD-V2 MEDIA (81Gi) (gizmo volume replication)
Backup/restore STAGING DB (gizmo copy) (gizmo volume replication)
Backup/restore STAGING MEDIA (217Gi) (gizmo volume replication)

Deploy Discourse PROD
Deploy Discourse STAGING

Recover PROD Discourse
Recover STAGING Discourse

Restore DB backup processes
  prod
  v2-prod
  staging (new)
Restore media & DB replicator
  prod

Move DM DB
Move AS DB

Deploy Job Operator
Deploy Jupyter Operator
Deploy AS
Deploy DM
Deploy UI

Adjust fragalysis stack GitHub CI Acrtion
  Needs to use the latest trigger implementation
  
Test CI process

Ensure S3 backups are operating

Scale down developer applications
Remove developer volumes
Remove developer cluster nodes

Scale down production applications
Remove Production volumes
Remove production cluster nodes

Reassess node resources
  Do we need more or less machines or a size change?

## Application sizing estimate
- Stacks need 15Gi RAM - so nodes need to provide at least that
- Media vol size 400Gi (each)
- DB vol size 50Gi (each)

**Comment by mwinokan** [comment]
@alanbchristie has migrated the stacks, database and graph, and is working on finalising the CI: https://github.com/xchem/fragalysis-stack/pull/104

**Comment by mwinokan** [comment]
@alanbchristie reports that the migration is complete but the old hardware is still being kept around

@alanbchristie please try copying the production data onto the staging stack to test, and if the migrations fail please wipe staging

**Comment by mwinokan** [comment]
@phraenquex tells @alanbchristie to kill the old cluster

**Comment by alanbchristie** [comment]
The DEV, PROD and Rancher cluster have been removed. Migration to Kubernetes 1.30 is complete. Looking at the timesheet it looks like migration effort was **98.25 hours** (12.3 days at 8 hours per day?).
